core/src/main/java/com/linecorp/armeria/client/endpoint/RingHashEndpointSelectionStrategy.java

```diff
@@ -0,0 +1,205 @@
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.XxHash;
+
+import it.unimi.dsi.fastutil.ints.Int2ObjectAVLTreeMap;
+
+import it.unimi.dsi.fastutil.ints.Int2ObjectSortedMap;
+
+final class RingHashEndpointSelectionStrategy implements EndpointSelectionStrategy{
+
+    static final RingHashEndpointSelectionStrategy INSTANCE = new RingHashEndpointSelectionStrategy();
+
+    private RingHashEndpointSelectionStrategy() {}
+
+    @Override
+    public EndpointSelector newSelector(EndpointGroup endpointGroup) {
+        return new RingHashSelector(endpointGroup);
+    }
+    
+    static class RingHashSelector extends AbstractEndpointSelector {
+
+        @Nullable
+        private volatile WeightedRingEndpoint weightedRingEndpoint;
+
+        RingHashSelector(EndpointGroup endpointGroup) {
+            super(endpointGroup);
+            endpointGroup.addListener(endpoints ->
+                weightedRingEndpoint = new WeightedRingEndpoint(endpoints), true
+            );
+        }
+
+        @Override
+        public Endpoint selectNow(ClientRequestContext ctx) {
+            final WeightedRingEndpoint weightedRingEndpoint = this.weightedRingEndpoint;
+            if (weightedRingEndpoint == null) {
+                // 'endpointGroup' has not been initialized yet.
+                return null;
+            }
+            return weightedRingEndpoint.select(ctx.endpoint());
+        }
+        
+        private final class WeightedRingEndpoint {
+            private final Int2ObjectSortedMap<Endpoint> ring = new Int2ObjectAVLTreeMap<>();
+            private final Iterable<Endpoint> endpoints;
+
+            Endpoint select(Endpoint point) {
+                final int key = getXXHash(point.host());
+                final SortedMap<Integer, Endpoint> tailMap = ring.tailMap(key);
+                return tailMap.isEmpty() ? ring.get(ring.firstKey()) : tailMap.get(tailMap.firstKey());
+            }
+
+            WeightedRingEndpoint(Iterable<Endpoint> endpoints) {
+                // prepare immutable endpoints
+                this.endpoints = Streams.stream(endpoints)
+                                        .filter(e -> e.weight() > 0) // only process endpoint with weight > 0
+                                        .sorted(Comparator.comparing(Endpoint::weight)
+                                                          .thenComparing(Endpoint::host)
+                                                          .thenComparingInt(Endpoint::port))
+                                        .collect(toImmutableList());
+
+                // The GCD is properly sized so that it does not exceed the size of the ring
+                final int gcd = findGcdInEndpoints(this.endpoints);
+                final int numberOfEndpointInTheRing = caculateNumberOfEndpointInTheRing(this.endpoints, gcd);
+                final int sizeOfRing = getSize(this.endpoints);
```

Comment:
No reason to use a more abstract type for implementations. 
```suggestion
            WeightedRingEndpoint(List<Endpoint> endpoints) {
                // prepare immutable endpoints
                this.endpoints = endpoints.stream()
                                        .filter(e -> e.weight() > 0) // only process endpoint with weight > 0
                                        .sorted(Comparator.comparing(Endpoint::weight)
                                                          .thenComparing(Endpoint::host)
                                                          .thenComparingInt(Endpoint::port))
                                        .collect(toImmutableList());

                // The GCD is properly sized so that it does not exceed the size of the ring
                final int gcd = findGcdInEndpoints(this.endpoints);
                final int numberOfEndpointInTheRing = caculateNumberOfEndpointInTheRing(this.endpoints, gcd);
                final int sizeOfRing = this.endpoints.size();
```